Unit Tests for Util (#366)
pez-globo/pufferfish-vent-software: firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Containers/Array.txt

Scenario: Array: make_array function creates array correctly.
  GIVEN('The make_array function')
   WHEN('an array is created from 8 uint8_t values as input')
      THEN('the value at every index in an array is as expected (0x0f, 0x36, 0x08, 0xff, 0x89, 0x01, 0x80, 0x02)')
      THEN('Size of an array is as expected 8')

  GIVEN('The make_array function')
   WHEN('an array is created from 3 strings as input parameter')
     THEN('the value of an array at every index  is as expected ("ab", "bc", "cd")')
     THEN('Size of an array is as expected 3')

  GIVEN('The make_array function') 
  WHEN('an array is created from const references of values, 0x02 and 0x0f')
    THEN('the array is as expected at every index [0x02,0x0f]')

  WHEN('an array is created from const references, value of reference variable is changed to -0x0c and 0x0a')
    THEN('the array is as expected at every index [0x0c,0x0a]')

>     THEN('Returns correct changed values for reference variables [0x0c, 0x0a]')

Reviewer: I think it would be clearer to construct these tests like so:

- Given the make_array function
  - When an array is created of const references to two variables with values 0x02 and 0x0f
    - Then the array has the expected values, 0x02 and 0x0f
    - Then changing the value of the first referenced variable to 0x0c also changes the value at the first index of the array to 0x0c
    - (etc.)

  WHEN('an array is created from non- const references of values ,0x00 and 0x0c')
    THEN('the array is as expected at every index [0x00,0x0c]')

  WHEN('an array is created from non- const references, values of referenced variables changed to ,0x02 and 0x86')
    THEN('the array is as expected at every index [0x02,0x86]')
    THEN('Returns correct changed values for reference variables [0x02, 0x86]')
    

> Scenario: Empty an array

Reviewer: This scenario is not about making an array empty.

  GIVEN('The make_array function')

>   WHEN('function empty is used on array (0x02, 0x03, 0x0a) ')

Reviewer: How are we using the `make_array` function in this test? I don't see a reference to making an array using `make_array`. If you're doing what I think you're doing, perhaps you could just include "Then the empty method of the array returns false" as a THEN in each of the above WHENs.

> Scenario: Util:Array make_array function works correctly for uint8, uint16, uint32, and int16 types

Reviewer: [coverage & traceability] It would be better if you also tested a `struct` here for basic functionality. The next scenario tests structs, but it also tests `std::reference_wrapper`s, i.e. two things are changed together and we don't have a test where we only change one thing yet. I think it's reasonable to expect that if `make_array` works on one primitive type (such as a `uint`) then it will work on all of them, especially since `make_array` is unrelated to the specific bytes inside the type. It is not necessarily obvious that `make_array` works on `struct`s if we only have tests showing it works on ints. However, I will not block the PR on this if you don't want to do this.

      THEN('The array has expected values of uint16_t at every index ') 

    WHEN('An array is created from input type uint16_t as the template parameters and with 2 uint32_t as the input values') 
       THEN('The array has the expected values of uint16_t at every index') 

    WHEN('An array is created from input type uint32_t as the template parameters and with 2 uint8_t input values') 
      THEN('The array has expected values of  uint8_t at every index') 

    WHEN('An array is created from input type uint32_t as the template parameters and with 2 uint16_t input values') 
      THEN('The array has expected values of  uint16_t at every index') 

    WHEN('An array is created from input type uint32_t as the template parameters and with 2 uint32 input values ') 
      THEN('The array has the expected values of  uint32_t at every index') 

    WHEN('An array is created from input type int16_t as the template parameters and with 2 int16  input values') 
      THEN('The array has the expected values of  int16_t at every index') 

Scenario: Util:Array make_array function works properly for const and non-const references
  GIVEN('Util:Array make_array function') 

>     WHEN('an array is created from two elements of structure having values 0x02 and 0x0f')

Reviewer: [clarity] This should specify what kind of array we're making. We should be making an array of const references. Here's one way you could word this:

```suggestion
    WHEN('an array of const references is created from two struct instances with internal values of 0x67 and 0x3e') 
```

      THEN(' the array has the expected values, 0x02 and 0x0f') 

>       THEN('Element values of structure are changed to 0x03 and 0xde, values of array at index 0 and at index 1 also change to 0x03 and 0xde') 

Reviewer: [clarity] the wording should have a clearer cause-and-effect relationship. Right now it's ambiguous because you're just asserting two declarations ("Element values of structure are changed to 0x03 and 0xde" and also "values of array at index 0 and at index 1 also change to 0x03 and 0xde"): you could be asserting that when the array is created, it has the expected values and then it changes the elements of the structure on its own. Here's an alternative way you could write this:

```suggestion
      THEN('Changing the values in the structure to 0x03 and 0xde also changes values of array accessible at indices 0 and 1 to 0x03 and 0xde') 
```

>       THEN('Values of an array can't be changed as elements in array are const reference')

Reviewer: [implementability of test] If we don't have a way to write compilable test code, maybe we should comment out this THEN and add a comment above it (in this file and in the cpp file) explaining that the code trying to change the values of the array doesn't compile. I think it's also fine to just omit this THEN entirely.

>     WHEN('an array is created from two non const elements of structure with values 0x67 and 0x3e') 

Reviewer: [clarity] This should specify what kind of array we're making. We should be making an array of non-const references. The two structs themselves should not be const. Here's one way you could word this:

```suggestion
    WHEN('an array of non-const references is created from two struct instances with internal values of 0x67 and 0x3e') 
```

       
    WHEN('An array is created from input type uint16_t as the template parameters and with 2 uint16 as the input values') 
      THEN('The array has expected values of uint16_t at every index ') 

    WHEN('An array is created from input type uint16_t as the template parameters and with 2 uint32_t as the input values') 
       THEN('The array has the expected values of uint16_t at every index') 

    WHEN('An array is created from input type uint32_t as the template parameters and with 2 uint8_t input values') 
      THEN('The array has expected values of  uint8_t at every index') 

    WHEN('An array is created from input type uint32_t as the template parameters and with 2 uint16_t input values') 
      THEN('The array has expected values of  uint16_t at every index') 

    WHEN('An array is created from input type uint32_t as the template parameters and with 2 uint32 input values ') 
      THEN('The array has the expected values of  uint32_t at every index') 

    WHEN('An array is created from input type int16_t as the template parameters and with 2 int16  input values') 
      THEN('The array has the expected values of  int16_t at every index') 

>     WHEN('An array is created from two struct instances with internal values of 0xb9, 0xe2, 0xa7 and  0x31')

Reviewer: [proofreading] this should be four struct instances, right?

      THEN('the array has the expected values, 0xb9, 0xe2, 0xa7 and 0x31') 

Scenario: Util:Array make_array function works properly for const and non-const references
  GIVEN('Util:Array make_array function') 
    WHEN('an array of const references is created from two struct instances with internal values of 0x02 and 0x0f')     
      THEN(' the array has the expected values, 0x02 and 0x0f') 
      THEN('Changing the values in the structure to 0x03 and 0xde also changes values of array accessible at indices 0 and 1 to 0x03 and 0xde')     
// values of an array can't be changed, as values are cont references
    
    WHEN('an array of non-const references is created from two struct instances with internal values of 0x67 and 0x3e')       
      THEN('The array has the expected values, 0x67 and 0x3e') 

>       THEN('Values of non-const reference elements from structure are changed to 0xa3 and 0xee, values of array at index 0 and at index 1 changes to 0xa3 and 0xee') 

Reviewer: [clarity] the wording should have a clearer cause-and-effect relationship. Right now it's ambiguous because you're just asserting two declarations ("values ... are changed to ..." and also "values ... changes to ..."): you could be asserting that when the array is created, it has the expected values and then it changes the elements of the structure on its own.
